File: C/BinaryTree.c
#include <stdio.h>
#include <stdlib.h>

#define TRUE 1
#define FALSE 0

typedef int data;

typedef struct BTree {
    data value;
    struct BTree *left;
    struct BTree *right;
} btree;

btree *CreateBtree() {
    btree *root = (btree *) malloc(sizeof(btree));
    root->value = NULL;
    root->left = NULL;
    root->right = NULL;
    return root;
}

int isEmpty(btree *bt) {
    if (bt->value == NULL)
        return TRUE;
    else
        return FALSE;
}


void SetLeftSubTree(btree *parent, btree *LChild) {
    //자식이 이미 있으면 메모리 해제, 메모리 누수 가능성 존재
    if (parent->left != NULL)
        free(parent->left);
    parent->left = LChild;
}

void SetRightSubTree(btree *parent, btree *RChild) {
    //자식이 이미 있으면 메모리 해제, 메모리 누수 가능성 존재
    if (parent->right != NULL)
        free(parent->right);
    parent->right = RChild;
}

btree *GetLChild(btree *bt) {
    return bt->left;
}

btree *GetRChild(btree *bt) {
    return bt->right;
}

data GetData(btree *bt) {
    return bt->value;
}

void SetData(btree *bt, data item) {
    bt->value = item;
}

//중위 순회
void inOrderPrint(btree* bt){
    if(bt == NULL)
        return;

    inOrderPrint(GetLChild(bt));
    printf("%d ", GetData(bt));
    inOrderPrint(GetRChild(bt));

}

//전위 순회
void preOrderPrint(btree* bt){
    if(bt == NULL)
        return;
    printf("%d ", GetData(bt));
    preOrderPrint(GetLChild(bt));
    preOrderPrint(GetRChild(bt));
}

//후위 순회
void postOrderPrint(btree* bt){
    if(bt == NULL)
        return;

    postOrderPrint(GetLChild(bt));
    postOrderPrint(GetRChild(bt));
    printf("%d ", GetData(bt));
}

int main() {
    btree* bt1 = CreateBtree();
    btree* bt2 = CreateBtree();
    btree* bt3 = CreateBtree();
    btree* bt4 = CreateBtree();
    btree* bt5 = CreateBtree();
    btree* bt6 = CreateBtree();
    btree* bt7 = CreateBtree();

    SetData(bt1, 1);
    SetData(bt2, 2);
    SetData(bt3, 3);
    SetData(bt4, 4);
    SetData(bt5, 5);
    SetData(bt6, 6);
    SetData(bt7, 7);

    SetLeftSubTree(bt1, bt2);
    SetRightSubTree(bt1, bt3);
    SetLeftSubTree(bt2, bt4);
    SetRightSubTree(bt2, bt5);
    SetLeftSubTree(bt3, bt6);
    SetRightSubTree(bt3, bt7);


    inOrderPrint(bt1); //중위 순회
    printf("\n");
    preOrderPrint(bt1); //전위 순회
    printf("\n");
    postOrderPrint(bt1); //후위 순회
}

File: TS&JS/BinaryTree.js
var BinaryTree = /** @class */ (function () {
    function BinaryTree(data) {
        if (data === void 0) { data = NaN; }
        this.data = data;
    }
    BinaryTree.prototype.isEmpty = function () {
        if (this.data === NaN)
            return true;
        else
            return false;
    };
    BinaryTree.prototype.SetLeftSubTree = function (LChild) {
        this.left = LChild;
    };
    BinaryTree.prototype.SetRightSubTree = function (RChild) {
        this.right = RChild;
    };
    BinaryTree.prototype.GetLChild = function () {
        return this.left;
    };
    BinaryTree.prototype.GetRChild = function () {
        return this.right;
    };
    BinaryTree.prototype.GetData = function () {
        return this.data;
    };
    BinaryTree.prototype.SetData = function (data) {
        this.data = data;
    };
    //중위 순회
    BinaryTree.prototype.inOrderPrint = function () {
        if (this.GetLChild() !== undefined) {
            this.GetLChild().inOrderPrint();
        }
        process.stdout.write(this.GetData() + " ");
        if (this.GetRChild() !== undefined) {
            this.GetRChild().inOrderPrint();
        }
    };
    //전위 순회
    BinaryTree.prototype.preOrderPrint = function () {
        process.stdout.write(this.GetData() + " ");
        if (this.GetLChild() !== undefined) {
            this.GetLChild().preOrderPrint();
        }
        if (this.GetRChild() !== undefined) {
            this.GetRChild().preOrderPrint();
        }
    };
    //후위 순회
    BinaryTree.prototype.postOrderPrint = function () {
        if (this.GetLChild() !== undefined) {
            this.GetLChild().postOrderPrint();
        }
        if (this.GetRChild() !== undefined) {
            this.GetRChild().postOrderPrint();
        }
        process.stdout.write(this.GetData() + " ");
    };
    return BinaryTree;
}());
var bt1 = new BinaryTree(1);
var bt2 = new BinaryTree(2);
var bt3 = new BinaryTree(3);
var bt4 = new BinaryTree(4);
var bt5 = new BinaryTree(5);
var bt6 = new BinaryTree(6);
var bt7 = new BinaryTree(7);
bt1.SetLeftSubTree(bt2);
bt1.SetRightSubTree(bt3);
bt2.SetLeftSubTree(bt4);
bt2.SetRightSubTree(bt5);
bt3.SetLeftSubTree(bt6);
bt3.SetRightSubTree(bt7);
bt1.inOrderPrint();
console.log("\n");
bt1.postOrderPrint();
console.log("\n");
bt1.preOrderPrint();
